### PR TITLE
Shladkyi - add comparison tests

### PR DESCRIPTION
**Pull request by SerhiiHladkyi**

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/rozetka/components/ComparisonPopup.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import com.softserveinc.ita.rozetka.page_objects.ComparisonPage;
+import io.qameta.allure.Step;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static java.lang.String.*;
+
+public class ComparisonPopup {
+
+    @Step("Comparison popup: open comparison page by number {number}")
+    public ComparisonPage openComparisonPageByNumber(int number) {
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/components/ComparisonPopup.java]: openComparisonPageFromList ?

File: src/main/java/com/softserveinc/ita/rozetka/components/Header.java [hunk 2]
```diff
@@ -37,4 +37,10 @@ public HomePage openHomePageByLogo() {
         $x("//*[@class='header__logo']").click();
         return new HomePage();
     }
+
+    @Step("Header: get comparison popup")
+    public ComparisonPopup getComparisonPopup() {
+        $x("//rz-comparison//button[@class='header__button ng-star-inserted']").click();
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/components/Header.java:43]: re-do

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+public class ComparisonPage extends BasePage {
+
+    public List<String> getCharacteristicList() {
+        return $$x("//dd[@class='comparison-characteristic__value']")
+                .shouldHave(sizeGreaterThanOrEqual(1))
+                .texts();
+    }
+
+    @Step("Comparison page: open different characteristics")
+    public ComparisonPage openDifferentCharacteristics() {
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: showDifferences()

File: src/test/java/com/softserveinc/ita/rozetka_test/ComparisonTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Test
+    public void testComparisonSettings() {
+        var allCharacteristics = comparisonPage.getCharacteristicList();
+        var firstProductCharacteristics = new ArrayList<>();
+        var secondProductCharacteristics = new ArrayList<>();
+
+        range(0, allCharacteristics.size())
+                .forEach(index -> (index % 2 == 0 ? firstProductCharacteristics : secondProductCharacteristics)
+                        .add(allCharacteristics.get(index)));
```
Comment (toleksyn) [inline src/test/java/com/softserveinc/ita/rozetka_test/ComparisonTest.java]: comparisonPage.getCharacteristicForProduct(1)

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        range(0, allCharacteristics.size())
+                .forEach(index -> (index % 2 == 0 ? firstProductCharacteristics : secondProductCharacteristics)
+                        .add(allCharacteristics.get(index)));
+        switch (number) {
+            case 1:
+                return firstProductCharacteristics;
+            case 2:
+                return secondProductCharacteristics;
+            default:
+                throw new IllegalArgumentException();
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: re-do

File: src/test/java/com/softserveinc/ita/rozetka_test/ComparisonTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                .anyMatch(index -> firstProductCharacteristics
+                        .get(index)
+                        .equals(secondProductCharacteristics.get(index)));
+        assertTrue(isCharacteristicHaveSameItem, "Characteristics of two products should have same item");
+
+        comparisonPage.showDifferences();
+        var differentFirstProductCharacteristics = comparisonPage.getCharacteristicForProduct(1);
+        var differentSecondProductCharacteristics = comparisonPage.getCharacteristicForProduct(2);
+
+        var isAllCharacteristicsDifferent = range(0, differentFirstProductCharacteristics.size())
```
Comment (toleksyn) [inline src/test/java/com/softserveinc/ita/rozetka_test/ComparisonTest.java]: are*

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java (hunk 5 above)
Comment (SerhiiHladkyi) [reply src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: fix

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+        switch (number) {
+            case 1:
+                return iterate(0, index -> index < allCharacteristics.size(), index -> index + 2)
+                        .mapToObj(allCharacteristics::get)
+                        .collect(toCollection(ArrayList::new));
+            case 2:
+                return iterate(1, index -> index < allCharacteristics.size(), index -> index + 2)
+                        .mapToObj(allCharacteristics::get)
+                        .collect(toCollection(ArrayList::new));
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: duplication, re-do

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            default:
+                throw new IllegalArgumentException();
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: no empty exceptions

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        if (number == 1) {
+            stream = iterate(0, index -> index < allCharacteristics.size(), index -> index + 2);
+        } else if (number == 2) {
+            stream = iterate(1, index -> index < allCharacteristics.size(), index -> index + 2);
+        }
+        if (number == 1 || number == 2) {
+            return stream
+                    .mapToObj(allCharacteristics::get)
+                    .collect(toCollection(ArrayList::new));
+        } throw new IllegalArgumentException("Incorrect number of product");
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: brackets

File: src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        IntStream stream = null;
+        if (number == 1) {
+            stream = iterate(0, index -> index < allCharacteristics.size(), index -> index + 2);
+        } else if (number == 2) {
+            stream = iterate(1, index -> index < allCharacteristics.size(), index -> index + 2);
+        }
+        if (number == 1 || number == 2) {
+            return stream
+                    .mapToObj(allCharacteristics::get)
+                    .collect(toCollection(ArrayList::new));
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/rozetka/page_objects/ComparisonPage.java]: collect(toList())

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Pull request merged**